### PR TITLE
Import configurator

### PR DESCRIPTION
**Pull request by rgrinberg**
Importing configurator into jbuilder (as discussed).

- [x] remove base
- [x] make bootstrap aware of configurator
- [ ] version the api
- [x] change the package to `dune.configurator`.

To consider:

* Should I bring back the various sexp functions? These didn't seem so useful and I should probably use the jbuilder combinator API to rewrite them.

### CODE REVIEW

**Comment by diml** [comment]
Thanks. Before we make the configurator library public, we need look at how to integrate it better with jbuilder, and in particular what the design will look like regarding the new stanza, etc... I expect that this might change how the .ml files are written and how parameters are passed from jbuild files to the configuration programs.

Additionally, we shouldn't make the jbuilder library public yet. The API is changing too much and too quickly. It seems fine to use it in configurator, but then we should hide the API to be safe, i.e. resuscitate the `public_interfaces` feature (#106). #427 should fix the issues of the original PR, and it has been implemented in #430. 

Why does the bootstrap need to be aware of configurator BTW?

**Comment by rgrinberg** [comment]
> Thanks. Before we make the configurator library public, we need look at how to integrate it better with jbuilder, and in particular what the design will look like regarding the new stanza, etc... I expect that this might change how the .ml files are written and how parameters are passed from jbuild files to the configuration programs.

Sure. There's no rush. We can have a design discussion or if you have an implementation in mind, feel free to push commits here as well.

> Additionally, we shouldn't make the jbuilder library public yet. The API is changing too much and too quickly. It seems fine to use it in configurator, but then we should hide the API to be safe, i.e. resuscitate the public_interfaces feature (#106). #427 should fix the issues of the original PR, and it has been implemented in #430.

OK, sure. Though I don't think this is a blocking issue. I don't think we've promised API stability of anything apart from configurator to anyone. But in practice #430 is likely to land soon anyway, so it won't be an issue.

> Why does the bootstrap need to be aware of configurator BTW?

My mistake, it doesn't. Bootstrapping needed a little fix, everything is working now.

**Comment by diml** [comment]
> Sure. There's no rush. We can have a design discussion or if you have an implementation in mind, feel free to push commits here as well.

I haven't thought about it much, I though you might have had a design in mind :)

> OK, sure. Though I don't think this is a blocking issue. I don't think we've promised API stability of anything apart from configurator to anyone. But in practice #430 is likely to land soon anyway, so it won't be an issue.

Hmm, I tend to disagree. I don't think we should make any release of jbuilder where one can access the internal APIs of jbuilder before we are happy to make them public. Otherwise it's like the compiler libraries: there is no stability guarantee but everybody is using them and the constant breakages are a huge pain.

**Comment by rgrinberg** [comment]
OK, I fixed the build on 4.02.3 as well. I'll wait for profiles (or custom variables) for a good way to disable the unattached doc warnings for Configurator.mli on 4.02.3

**Comment by diml** [comment]
Having easy support for pkg-config would be great. I think it should be a field of libraries, since that where we put C-stubs.

**Comment by rgrinberg** [comment]
Listing a couple of other requirements for pkg config support:

* Support OPAM_PKG_CONFIG_PATH as a way to prepend to PKG_CONFIG. (https://github.com/mirage/io-page/blob/master/lib/config/discover.ml#L9)

* Make pkg-config in conjunction with various package mangers like macports (homebrew is already supported). I suppose this means having a smarter default for when pkg-config is absent.

I imagine the field would look like something like this:

```
(pkg-config
 ((libs (sqlite gsl)))
```

I've also looked at a few discover scripts and have noticed that:

* Searching for and parsing header files is fairly common. I suppose some of these use cases can be replaced by writing C that uses the headers and then using the preprocessor to extact the values. But some uses seem to rely on actually parsing C. For example, to get a list of constants defined in a header file. Parsing C is way too hard to help with, but we could provide a consistent API for searching for particular headers.

* Users rely on environment variables to toggle features in their configure scripts. Unfortunately, that means changes to these env vars don't trigger the rebuilds they should. We could address the problem very simply by adding environment variables as another thing you can depend (files, aliases, etc.). Or we could be more fancy and have a mechanism for configurator scripts to output the list of env vars they've read. I suppose that would be a little more inline with the proposed mechanism for pp's outputting dependencies.

* Some scripts also parse the flags given by pkg-config. Having a better API for doing this could be useful.

Otherwise, the current configurator API seems fairly good to me. I'd be interested to hear any feature requests from current users of configurator: @Chris00 @mmottl @avsm @samoht

**Comment by mmottl** [comment]
@rgrinberg Thanks for your work on this.  I'm pretty happy with the current state already and haven't run into any issues for quite a while.  Anything that makes configurator files even shorter and simpler is welcome.  `gsl` in particular has to read header files for code generation and depends on environment variables.  If you find generic solutions that help with that, I'd be happy to try them out.

**Comment by diml** [comment]
I think one thing we could improve is how to pass values from the jbuild file to the configurator script. Having to parse the command line arguments in the script is annoying. We could imagine writing something like this in the jbuild file:

```scheme
(x "${version:foo}")
(y ${lib-available:bar})
```

that would generate:

```ocaml
let x = "42"
let y = false
```

BTW, we could also dump the OCaml configuration this way, so that configurator doesn't have to call `ocamlc -config` again.

Another thing I have been thinking is that when you have a lot of tests, such as in core, then the configurator program takes a bit of time to execute, which is a bit sad since these tests could be run in parallel. If we mange to cut the boilerplate to a minimum, then it should be OK to split these into one test per .ml and have a final .ml that merge the results.

**Comment by rgrinberg** [comment]
@diml what do you think of merging this library but not making the library public? Will save us a lot of conflict resolution.

**Comment by diml** [comment]
Ok, sounds good. I'm doing my review now

File: bootstrap.ml [hunk 1]
```diff
@@ -36,6 +36,7 @@ let dirs =
   ; "src/xdg"                , Some "Xdg"
   ; "vendor/boot"            , None
   ; "src/usexp"              , Some "Usexp"
+  ; "src/usexp"              , Some "Configurator"
```
Comment (diml) [inline bootstrap.ml]: Why do we need this?

File: vendor/boot/jbuilder_opam_file_format.ml [hunk 2]
```diff
@@ -16,7 +16,7 @@ end
 module OpamBaseParser = struct
   open OpamParserTypes
   let main _lex _lexbuf fn =
-    assert (fn = "jbuilder.opam");
+    assert (fn = "jbuilder.opam" || fn = "configurator.opam");
```
Comment (diml) [inline vendor/boot/jbuilder_opam_file_format.ml]: I suppose this is stale?

**Comment by rgrinberg** [comment]
I also need to get rid of all the `(package jbuilder)` stuff.

**Comment by rgrinberg** [comment]
@diml before merging, I'd like to add a test. so let me know once you're happy with this. The test will have to be a unit test as we haven't yet figured out a good way to make our private libraries available within the blackbox sandbox.

**Comment by diml** [comment]
OK, I'm happy with the current state.

Eventually I'd like that we don't depend on the jbuilder library and that  instead we depend on stdune, fiber and a new process library we need to create by extracting Process and Scheduler. And probably and Ocamlc_config one as well

**Comment by rgrinberg** [comment]
I've rewritten the history. In particular, to maintain all the changes done to the original copy of configurator as part of the import. Will merge once CI passes.

**Pull request merged**